nsv13/code/modules/munitions/ship_weapons/_ship_weapon.dm

```diff
@@ -0,0 +1,470 @@
+	icon = 'nsv13/icons/obj/railgun.dmi'
+	icon_state = "OBC"
+	density = TRUE
+	anchored = TRUE
+	layer = BELOW_OBJ_LAYER
+
+	var/obj/structure/overmap/linked = null
+	var/obj/machinery/computer/ship/munitions_computer/linked_computer
+	var/obj/weapon_overlay/overlay = null
+	var/list/icon_state_list
+
+	// Icons, sounds, and timing for the states
+	var/load_sound = 'nsv13/sound/effects/ship/mac_load.ogg'
+	var/mag_load_sound = 'sound/weapons/autoguninsert.ogg'
+	var/load_delay = 20
+
+	var/unload_sound = 'nsv13/sound/effects/ship/freespace2/crane_short.ogg'
+	var/mag_unload_sound = 'sound/weapons/autoguninsert.ogg'
+	var/unload_delay = 10
+
+	var/feeding_sound = 'nsv13/sound/effects/ship/freespace2/crane_short.ogg'
+	var/fed_sound = 'nsv13/sound/effects/ship/reload.ogg'
+	var/feed_delay = 20
+
+	var/chamber_sound = 'nsv13/sound/weapons/railgun/ready.ogg'
+	var/chamber_delay_rapid = 2
+	var/chamber_delay = 10
+
+	var/firing_sound = 'nsv13/sound/effects/ship/mac_fire.ogg'
+	var/fire_animation_length = 5
+	var/fire_mode
+
+	var/malfunction_sound = 'sound/effects/alert.ogg'
+
+	//Various traits that probably won't change
+	var/maintainable = TRUE //Does the weapon require maintenance?
+	var/bang = TRUE //Is firing loud?
+	var/auto_load = FALSE //Does the weapon feed and chamber the round once we load it?
+	var/semi_auto = FALSE //Does the weapon re-chamber for us after firing?
+
+	var/ammo_type = null
+	var/magazine_type = null
+	var/max_ammo = 1
+
+	var/datum/ship_weapon/weapon_type
+
+	// Things that change while we're operating
+	var/maint_req = 0 //Number of times a weapon can fire until a maintenance cycle is required. This will countdown to 0.
+	var/malfunction = FALSE
+	var/maint_state = MSTATE_CLOSED
+	var/safety = TRUE //Can only fire when safeties are off
+	var/loading = FALSE
+	var/state = STATE_NOTLOADED
+
+	var/obj/item/ammo_box/magazine/magazine //Magazine if we have one
+	var/obj/chambered //Chambered round if we have one. Extrapolate ammo type from this
+	var/list/ammo = list() //All loaded ammo
+
+/**
+ * Constructor for /obj/machinery/ship_weapon
+
+ * Attempts to link the weapon to an overmap ship.
+ * If the weapon requires maintenance, generates initial maintenance countdown.
+ * Caches icon state list for sanity checking when updating icons.
+ */
+/obj/machinery/ship_weapon/Initialize()
+	. = ..()
+	get_ship()
+	if(maintainable)
+		maint_req = rand(15,25) //Setting initial number of cycles until maintenance is required
+		create_reagents(50)
+	icon_state_list = icon_states(icon)
+
+/**
+ * Destructor for /obj/machinery/ship_weapon
+ * Try to unlink from a munitions computer, so it can re-link to other things
+ */
+/obj/machinery/ship_weapon/Destroy()
+	. = ..()
+	if(linked_computer)
+		linked_computer.SW = null
+
+/**
+ * Tries to link the ship to an overmap by finding the overmap linked it the area we are in.
+ */
+/obj/machinery/ship_weapon/proc/get_ship()
+	var/area/AR = get_area(src)
+	if(AR.linked_overmap)
```

Comment:
we should probably have a world log to report that it failed to link as this is a good way to notice if something isnt setup properly on world loadup. (Could be annoying if we plan to put these in ruins or something though. Ask Kmc for what would be preferred maybe. If the former is preferred but not wanted to occur post roundstart there is a way to place it behind a define that gets removed on round start I believe)

Reply:
The only issue though is how many places call get_ship which is for a good reason. I think a better handle for this might need to be worked on first. So you can likely ignore my previous post if you wish.